### ISSUE TITLE
Allow having multiple conftest.py on same level

### ISSUE THREAD

**Issue by ikonst** [pytest-dev/pytest#3582]
In our project, the conftest.py files have grown so large that we felt the need to refactor them into smaller modules. However, modules imported by conftest.py using `import` cannot contribute fixtures.

We've tried using `pytest_plugin` but unfortunately (as the documentation says and as my testing confirms), importing a plugin with fixtures in one conftest.py makes them available in tests in parent and sibling hierarchies.

I'd like to propose a simple solution, to have pytest look both for `conftest.py and `conftest_*.py` under the same directory. This way, unruly conftest.py files could be easily split up.

(Let me know if it's an acceptable solution, I can implement it.)

**Comment by nicoddemus** [comment]
Hi @ikonst,

For your use case, where you want fixtures in separate files just as a means to keep things more manageable, you can import the fixtures normally into the `conftest.py` file:

```
# conftest.py
from ._app_fixture import app
from ._regression_fixture import regression
```

And so on. Importing fixtures into conftest files *in general* is not recommended as it might lead to subtle problems, but importing them into the `conftest.py` file (and nowhere else) looks like it will solve your problem.

**Comment by ikonst** [comment]
Would that force me to import fixtures one by one, though?

e.g. let's say I have 100 fixtures: 50 are data model fixtures (that I'd like to store in "conftest_data_models.py") and 50 are network call fixtures ("conftest_network_calls.py").

**Comment by nicoddemus** [comment]
No, you can import them using `*`; as long as they are in the module's namespace by the time pytest looks for fixtures, it will work.

**Comment by blueyed** [comment]
I wonder if it would be feasible or possible already to have a module "conftest"?
I.e. `conftest/__init__.py` and several other files in `conftest/`, that get imported in `__init__.py`.

**Comment by nicoddemus** [comment]
Currently we look explicitly for `conftest.py` files:

https://github.com/pytest-dev/pytest/blob/4d0297b4138b694e1d99f4c8147aaf06d725d0f4/src/_pytest/config.py#L367-L370

As to change this to support packages, it is possible I guess, but it would mean to change other places that also look for `conftest.py` files, for example:

https://github.com/pytest-dev/pytest/blob/4d0297b4138b694e1d99f4c8147aaf06d725d0f4/src/_pytest/doctest.py#L354-L355

https://github.com/pytest-dev/pytest/blob/4d0297b4138b694e1d99f4c8147aaf06d725d0f4/src/_pytest/fixtures.py#L1054-L1057

Because of the possibility of introducing more bugs, I'm not sure it is worth trying to add support for this given that the "workaround" is pretty simple: import your symbols into the `conftest.py` file, which you would need to do anyway with the `conftest/__init__.py` approach.

**Comment by RonnyPfannschmidt** [comment]
im -1 on this

**Comment by nicoddemus** [comment]
Yep, I agree with @RonnyPfannschmidt here.

@ikonst did importing the fixtures into the `conftest.py` fixed your original issue?

**Comment by ikonst** [comment]
Thanks for all your help. I'll test this on my codebase and report back.

**Comment by nicoddemus** [comment]
Thanks!

**Comment by ikonst** [comment]
`from .fixtures.models import *` worked out great

**Comment by ikonst** [comment]
To follow up, my approach to import all fixtures under `fixtures/*` is adding to conftest.py:
```py
from pkgutil import walk_packages

from . import fixtures

for package in walk_packages(fixtures.__path__, prefix=fixtures.__name__ + '.'):
    module = package.module_finder.find_module(package.name).load_module()
    globals().update({k: v for k, v in vars(module).items() if not k.startswith('_')})
```

Having to copy-paste this code snippet across multiple conftest.py files seems suboptimal, but so does implicitly specifying all fixtures, e.g.
```
from .fixtures.foo import *  # noqa
from .fixtures.bar import *  # noqa
```
and the `# noqa`s are somewhat noisy.

Furthermore, I didn't find a way to write a pytest plugin to hook into loading of conftest.py files, in order to inject this functionality.

**Comment by tpvasconcelos** [comment]
@nicoddemus any suggestions on how to achieve this when using the recommended `importlib` import mode? It seems to be impossible to perform any imports since `tests/` is not recognised as a package and isn't inserted in the sys.path like in the default `prepend` import mode.

Any suggestions are welcome here! :)

**Comment by tpvasconcelos** [comment]
For reference, I opted for using the [pythonpath](https://docs.pytest.org/en/7.4.x/reference/reference.html#confval-pythonpath) option to point to a separate directory containing helpers/utilities for the test suite.

---

As an example, 

1. Add an `extras/` directory to sys.path using the `pythonpath` option in the `pytest.ini` config file

    ```ini
    pythonpath = extras
    ```

2. Add your helper modules/package here:

    ```
    .
    ├── pytest.ini
    ├── extras/
    │   └── test_helpers/
    │       ├── helper_module_foo.py
    │       └── helper_module_bar.py
    ```

3. Impor them from your test modules as:

    ```python
    from test_helpers. helper_module_foo import foo
    from test_helpers. helper_module_bar import bar
    ```

4. Extra credit if you write tests for your test helpers.

The final project tree looks something like this for me:
```
.
├── pytest.ini
├── extras/
│   └── test_helpers/
│       ├── helper_module_foo.py
│       └── helper_module_bar.py
├── tests/
│   ├── conftest.py
│   ├── extras/
│   │   └── test_test_helpers/
│   │       ├── test_helper_module_foo.py
│   │       └── test_helper_module_foo.py
│   └── unit/
│       └── # my unit test suite goes here...
```

**Comment by danielschenk** [comment]
The `from .foo import *` workaround works fine for fixtures not starting with an underscore. However, in a project I'm working on we have some fixtures which we deem internal (eg. because they are intended to be purely `autouse`, or for use by another fixture in the same file only) which we start with an underscore to clarify this in a Pythonic way.

Since the star import only considers non-underscore items, this workaround does not feel great to me. I have to either do explicit imports or modify `__all__` to get this working, or just don't use underscores anymore in my fixture names.

I think that automatic fixture imports only working from test files or `conftest.py` is one of the biggest hassles for medium to large size pytest projects. `conftest.py` tends to grow out of control and become hard to read/navigate pretty quickly, and having to create subfolders just to split across multiple `conftest.py` files also isn't really nice (imagine having 5 `conftest.py` tabs open in your IDE...).

**Comment by The-Compiler** [comment]
@danielschenk It sounds like you just want to [use `pytest_plugins` instead](https://docs.pytest.org/en/8.0.x/how-to/plugins.html#available-installable-plugins), at least if this is in the root `conftest.py` file (which I assume it is).

**Comment by danielschenk** [comment]
> @danielschenk It sounds like you just want to [use `pytest_plugins` instead](https://docs.pytest.org/en/8.0.x/how-to/plugins.html#available-installable-plugins), at least if this is in the root `conftest.py` file (which I assume it is).

We will add sibling test folders in the near future which should be (largely) independent, so that is not going to work for me.
